### ISSUE TITLE
zmq_msg_close failed: Invalid message frame

### ISSUE THREAD

**Issue by 9il** [kyllingstad/zmqd#11]
Hi, Lars!
I have this error with example from README (both with 1.0.0-alpha and 0.5.2). Can you suggest something?

**Comment by kyllingstad** [comment]
Hi!

I cannot reproduce this (on Linux, DMD 2.067.1). I also find it very strange that you get this error. As far as I can tell it can only come from a failed assertion in `zmqd.Frame.~this()`, but the README example doesn't use `Frame` at all. Are you running a modified example?

Please paste the exact code that triggers the error, and provide me with some info about your environment (ZMQ version, compiler version, OS, etc.), and I will try to reproduce the error.

**Comment by 9il** [comment]
Looks like fails unittest in zmqd.
```
$ dub --version
DUB version 0.9.24-rc.2, built on Jul 26 2015

$ uname
Darwin

$ brew info zmq
zeromq: stable 4.1.2 (bottled), HEAD
High-performance, asynchronous messaging library
http://www.zeromq.org/
/usr/local/Cellar/zeromq/4.1.2 (63 files, 3.3M) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/zeromq.rb
==> Dependencies
Build: pkg-config ✔
Optional: libpgm ✘, libsodium ✘, norm ✘

$ dub test zmqd
Building package zmqd in /Users/ilya/GitHub/zmqd/
Generating test runner configuration '__test__library__' for 'library' (library).
Performing "unittest" build using dmd for x86_64.
Building zeromq 4.0.5+commit.4.g14048f0 configuration "library"...
Building zmqd 1.0.0-alpha+commit.10.g662e516 configuration "__test__library__"...
Linking...
Running ../../GitHub/zmqd/__test__library__ 
core.exception.AssertError@../../GitHub/zmqd/src/zmqd.d(2005): zmq_msg_close failed: Invalid message frame
----------------
5   __test__library__                   0x0000000105c078a2 nothrow @safe void zmqd.Frame.__dtor() + 98
6   __test__library__                   0x0000000105c026eb @safe void zmqd.Socket.__unittestL654_15() + 299
7   __test__library__                   0x0000000105be9262 void zmqd.__modtest() + 74
8   __test__library__                   0x0000000105c36a00 runModuleUnitTests + 256
9   __test__library__                   0x0000000105c323ee int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*)) + 38
10  __test__library__                   0x0000000105c4997d int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_osx.SectionGroup) + 81
11  __test__library__                   0x0000000105c4990d int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))) + 49
12  __test__library__                   0x0000000105c323c2 int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)) + 34
13  __test__library__                   0x0000000105c36985 runModuleUnitTests + 133
14  __test__library__                   0x0000000105c44b5b void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() + 23
15  __test__library__                   0x0000000105c44b11 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 45
16  __test__library__                   0x0000000105c44a84 _d_run_main + 504
17  __test__library__                   0x0000000105be9212 main + 34
18  libdyld.dylib                       0x00007fff9a44c5c9 start + 1
19  ???                                 0x0000000000000001 0x0 + 1
 
```

**Comment by 9il** [comment]
Fails on this:
```D
    unittest
    {
        // Sender
        auto snd = Socket(SocketType.req);
        snd.connect("inproc://zmqd_msg_receive_example");
        snd.send("Hello World!");

        // Receiver
        import std.string: representation;
        auto rcv = Socket(SocketType.rep);
        rcv.bind("inproc://zmqd_msg_receive_example");
        auto msg = Frame();
        rcv.receive(msg);
        assert (msg.data.asString() == "Hello World!");
    }
```

**Comment by kyllingstad** [comment]
All unittests pass without problem on my machine (Linux, ZMQ 4.0.5). Maybe it's a ZMQ 4.1 thing?

**Comment by 9il** [comment]
Thanks! _The sizeof(zmq_msg_t) changed from 32 bytes to 64 bytes._
I will open PR in ZeroMQ deimos.

**Comment by 9il** [comment]
PR https://github.com/D-Programming-Deimos/ZeroMQ/pull/19

**Comment by kyllingstad** [comment]
Now that ZMQ 4.1 is stable, I will update both deimos/zeromq and zmqd as soon as I find the time for it.

**Comment by kyllingstad** [comment]
D-Programming-Deimos/ZeroMQ#20

ZMQD update will follow shortly after this has been merged.

**Comment by 9il** [comment]
It was very fast! Whoohoo! Thanks!

**Comment by kyllingstad** [comment]
This should help catch similar errors in the future: 96d5f519a1ad17abc96d153480bb2c0da8671cb8

**Comment by kyllingstad** [comment]
D-Programming-Deimos/ZeroMQ has now been updated and tagged with `v4.1.2`. As soon as the Dub registry refreshes its package info, you should be able to use that version of ZeroMQ with ZMQD. Please reopen this bug report if it still doesn't work.

**Comment by 9il** [comment]
It does not work. `zmq_curve_keypair` defined in `zmq.d`, not `utils.d`.
I have not privileges to reopen this bug.
```
dub test
Fetching zeromq 4.1.2 (getting selected version)...
Placing zeromq 4.1.2 to /Users/ilya/.dub/packages/...
Generating test runner configuration '__test__library__' for 'library' (library).
Performing "unittest" build using dmd for x86_64.
Building zeromq 4.1.2 configuration "library"...
Building zmqd 1.0.0-beta+commit.1.g8afbd61 configuration "__test__library__"...
src/zmqd.d(2991,12): Error: module deimos.zmq.utils import 'zmq_curve_keypair' not found
dmd failed with exit code 1.
```

**Comment by kyllingstad** [comment]
Oops, sorry. Fixed now: ea4855727f348a399949dd0a475d1db15edfa830

**Comment by 9il** [comment]
Thanks, works now!
